### PR TITLE
integrated ckbtc and ckusdc

### LINKED ISSUE #503: Receive/Support BTC, USDC
Today we list the DOLR tokens holded by the principal in the Wallet.
We would want to increase its scope and show BTC and USDC tokens as well held by the users.

**Ask from dev**
UI Task in scope
Add a index ID row in the token details page

Backend task:
Support for any other token like NNS wallet does. If as a user, transfer any SNS token to my Yral Idex ID it should be visible in the Yral wallet.

This is inline to what NNS wallet.

Relies on index canister. We have canister ID, root also and ledger.

### PR DESCRIPTION
**Pull request by joel-medicala-yral**
closes #503

### CODE REVIEW

**Comment by harshita-srivastava-yral** [comment]
Working fine! Checked USDC however BTC I couldn't check due to insufficient funds. However this is good to go...! @joel-medicala-yral

File: ssr/src/page/token/mod.rs [hunk 1]
```diff
@@ -7,16 +7,15 @@ mod sns_form;
...
 pub mod types;
 
-use candid::Principal;
 use leptos::Params;
 use leptos_router::Params;
 
 #[derive(Params, PartialEq, Clone)]
 struct TokenParams {
-    token_root: Principal,
+    token_root: String,
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/mod.rs:15]: Ensure that these are statically typed
Feel free to use an enum

File: ssr/src/consts/mod.rs [hunk 2]
```diff
@@ -72,3 +72,19 @@ pub mod google {
...
+pub static HARDCODED_TOKEN_IDS: phf::Map<&'static str, HardCodedIDs> = phf::phf_map! {
+    "btc" => HardCodedIDs {
+        ledger: "mxzaz-hqaaa-aaaar-qaada-cai",
+        index: "n5wcd-faaaa-aaaar-qaaea-cai",
+    },
+    "usdc" => HardCodedIDs {
+        ledger: "xevnm-gaaaa-aaaar-qafnq-cai",
+        index: "xrs4b-hiaaa-aaaar-qafoa-cai"
+    }
+};
```
Comment (rupansh-sekar-yral) [inline ssr/src/consts/mod.rs:90]: Is this something we can't do with a `match` block?

File: ssr/src/page/token/info.rs [hunk 3]
```diff
@@ -144,20 +152,21 @@ fn TokenInfoInner(
...
                 {if key_principal.is_some() {
-                    view! { <Transactions source=IndexOrLedger::Index(meta.index) key_principal symbol=meta.symbol.clone()/> }
+                    view! { <Transactions source=IndexOrLedger::Index(meta.index) key_principal symbol=meta.symbol.clone() decimals/> }
                 } else {
                     view! {
                         <Transactions
                             source=IndexOrLedger::Ledger(meta.ledger)
                             key_principal=None
                             symbol=meta.symbol.clone()
+                            decimals
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/info.rs:169]: Unrelated to this PR, but from what I can see key_principal depends on source?
You should update `IndexOrLedger::Index` variant so it takes key_principal as well.

File: ssr/src/page/token/transfer.rs [hunk 4]
```diff
@@ -117,6 +118,79 @@ async fn transfer_token_to_user_principal(
...
+    destination_principal: Principal,
+    ledger_canister: Principal,
+    amount: TokenBalance,
+) -> Result<(), ServerFnError> {
+    let cans = cans_wire.canisters().unwrap();
+    // let user_id = user_id.to_owned();
+    // let user_principal = user_id.sender()?;
+    // let agent = cans.agent.get_agent().await;
+    // let user_principal = agent.get_principal()?;
+    // log::debug!("user_principal: {:?}", user_principal.to_string());
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/transfer.rs:132]: is this something we will use in the future?

File: ssr/src/page/token/transfer.rs [hunk 5]
```diff
@@ -117,6 +118,79 @@ async fn transfer_token_to_user_principal(
...
+            fee: None,
+            from_subaccount: None,
+            to: Account {
+                owner: destination_principal,
+                subaccount: None,
+            },
+            created_at_time: None,
+        })
+        .await
+        .unwrap();
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/transfer.rs:147]: ?

File: ssr/src/page/token/transfer.rs [hunk 6]
```diff
@@ -117,6 +118,79 @@ async fn transfer_token_to_user_principal(
...
+    // let res = agent
+    //     .update(
+    //         &destination_canister,
+    //         "add_token",
+    //     )
+    //     .with_arg(candid::encode_one(root_canister).unwrap())
+    //     .call_and_wait()
+    //     .await
+    //     .unwrap();
+    // println!("add_token res: {:?}", res);
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/transfer.rs:189]: will we use this code?

File: ssr/src/page/token/transfer.rs [hunk 7]
```diff
@@ -134,8 +208,9 @@ fn FormError<V: 'static>(#[prop(into)] res: Signal<Result<V, String>>) -> impl I
 #[component]
 fn TokenTransferInner(
     cans: Canisters<true>,
-    root: Principal,
+    root: Option<Principal>,
     info: TokenMetadata,
+    param: String,
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/token/transfer.rs:213]: This should be done with an enum
root and params are mutually exclusive no?

File: ssr/src/page/token/transfer.rs (hunk 5 above)
Comment (rupansh-sekar-yral) [reply ssr/src/page/token/transfer.rs:147]: I just noticed that these problems exist in `transfer_token_to_user_principal` as well, feel free to fix that function as well

File: ssr/src/page/wallet/tokens.rs [hunk 8]
```diff
@@ -33,8 +30,15 @@ impl KeyedData for Principal {
...
+impl KeyedData for String {
+    type Key = String;
+
+    fn key(&self) -> Self::Key {
+        self.clone()
+    }
+}
 impl CursoredDataProvider for TokenRootList {
-    type Data = Principal;
+    type Data = String;
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/tokens.rs:41]: Strong typing!!

File: ssr/src/page/wallet/tokens.rs [hunk 9]
```diff
@@ -90,14 +123,32 @@ pub fn TokenViewFallback() -> impl IntoView {
...
+            if let Some(HardCodedIDs { ledger, index }) = HARDCODED_TOKEN_IDS.get(&token_root) {
+                get_ck_metadata(
+                    &cans,
+                    Some(user_principal),
+                    Principal::from_text(ledger).unwrap(),
+                    Principal::from_text(index).unwrap(),
+                )
+                .await
+                .unwrap()
+                .unwrap() // Map AgentError to ServerFnError
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/tokens.rs:142]: are we really doing the mapping? :P

File: ssr/src/page/wallet/tokens.rs [hunk 10]
```diff
@@ -132,7 +187,7 @@ pub fn TokenTile(user_principal: String, token_meta_data: TokenMetadata) -> impl
     view! {
         <div class="flex  w-full items-center h-16 rounded-xl border-2 border-neutral-700 bg-white/15 gap-1">
             <a
-                href=format!("/token/info/{root}/{user_principal}?airdrop_amt=100")
+                href=format!("/token/info/{}/{user_principal}?airdrop_amt=100",  root.map(|r| r.to_text()).unwrap_or(info.name.to_lowercase()))
```
Comment (rupansh-sekar-yral) [inline ssr/src/page/wallet/tokens.rs:190]: Can we add functions for these share links? Its really hard to follow which link does what

**Review by rupansh-sekar-yral — CHANGES_REQUESTED** [review]

File: ssr/src/consts/mod.rs (hunk 2 above)
Comment (joel-medicala-yral) [reply ssr/src/consts/mod.rs:90]: Yes but we're are required to match it multiple times in the codebase to infer the param... Like for example we could be using an actual token root or we way just pass in btc or usdc as the param so we need to match this many times so it'd be better to have a static map and then get it whenever we need it... Or prolly could use a function instead...

File: ssr/src/page/token/transfer.rs (hunk 4 above)
Comment (joel-medicala-yral) [reply ssr/src/page/token/transfer.rs:132]: i actually copied it from the transfer token function (which is above it) so im not sure if we need it or not...

File: ssr/src/page/wallet/tokens.rs (hunk 8 above)
Comment (joel-medicala-yral) [reply ssr/src/page/wallet/tokens.rs:41]: Yeah the reason why its a string its because the param could be a hardcoded type like "btc" and "usdc" which will be changed to ledger instead of root so it should be fixed...

File: ssr/src/page/wallet/tokens.rs (hunk 9 above)
Comment (joel-medicala-yral) [reply ssr/src/page/wallet/tokens.rs:142]: woops forgot the remove the comment :>

File: ssr/src/page/wallet/tokens.rs (hunk 10 above)
Comment (joel-medicala-yral) [reply ssr/src/page/wallet/tokens.rs:190]: Sure

File: ssr/src/consts/mod.rs (hunk 2 above)
Comment (rupansh-sekar-yral) [reply ssr/src/consts/mod.rs:90]: you can just use a function, the compiler will convert the code to a static lookup table internally (which is probably more efficient than  phf)

**Review by rupansh-sekar-yral — APPROVED** [review]
